Architecture refactor #2 (Bot v3)

jjaxu opened the pull request:
- Relocating response logic to `BotCommand` to increase customizability
- Cleaning up imports
- Tweaking UT's to reflect new changes

jjaxu committed:
Refactoring architecture once more
- To enable better handling of other types of responses other than sendMessage
- Adding intentional failed UT to verify CI

jjaxu committed:
Fixing CI

jjaxu committed:
Fixing UT
- intentional to observe CI

george-lim commented on commands/botcommand.py:
```diff
@@ -13,3 +16,33 @@ def skip_edited(self) -> bool:
         raise NotImplementedError("Method is not implemented, did you forget to override?")
+
+    def handle_normal_query(self, response):
+        if self.query.is_edited and not ALLOW_EDITED_MESSAGES:
```
Are you ever planning on allowing edited messages? If not, you should just remove this flag and if statement altogether. It's a bad idea to have flags that will NEVER change.

george-lim commented on commands/botcommand.py:
```diff
@@ -1,9 +1,12 @@
 class BotCommand(object):
-    def __init__(self, command: str):
+    def __init__(self, query):
         super().__init__()
```
I personally would remove `super().__init__()` because it's just an object. This makes me think that your BotCommand is a subclass of a custom object. I would also remove the explicit subclass object here as well for the same reason

george-lim commented on commands/botcommand.py:
```diff
@@ -13,3 +16,33 @@ def skip_edited(self) -> bool:
+        data = dict()
+
+        if response is None:
+            print("Response is None: nothing sent back")
```
I think you should consider using `logging` library instead of `print` everywhere. You can do stuff like `logging.warning("...")` to be more explicit.

jjaxu replied:
Good suggestion, I'll keep a note of this for future changes

george-lim replied:
Sure, this can be done in another PR

george-lim commented on commands/botcommand.py:
```diff
@@ -13,3 +16,33 @@ def skip_edited(self) -> bool:
+        if self.query.is_edited and not ALLOW_EDITED_MESSAGES:
+            return
+
+        data = dict()
```
Avoid using very simple non-descriptive variables like `data` please. I would prefer something like `send_message_json`.

george-lim commented on commands/botcommand.py:
```diff
@@ -13,3 +16,33 @@ def skip_edited(self) -> bool:
+            data[PARSE_MODE_KEY] = self.parse_mode
+
+        if self.other_params:
+            data.update(self.other_params)
```
I'm not a fan of having `other_params`. I would split this function into 2 functions:  `format_send_message_json` and `send_message`. Then, have subclass commands override `format_send_message_json` to add their own parameters (instead of storing `other_params` in the parent class, it shouldn't know anything about "extra" parameters).

A child class could do then do something like:

```
def handle_normal_query(self, response):
  send_message_json = super.format_send_message_json(...)
  # do stuff with new parameters
  super.send_message(send_message_json)
```

george-lim reviewed (changes requested):
Good start, have a couple things that I would do differently to make the code cleaner. I think you should solve the import problem in another PR, because there is a lot of duplicate imports (like for each command for example). I think you should use `logging` instead of print though.

jjaxu committed:
Tweaking according to suggestions
- removing ALLOW_EDITED_MESSAGES flag in config
- adding format_json_response method for better flow

george-lim reviewed (approved):
LGTM 👍
